fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/synonym/SynonymAnalyzerRegistryImpl.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * SynonymAnalyzer.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.synonym;
+
+import com.apple.foundationdb.record.RecordCoreArgumentException;
+import com.apple.foundationdb.record.logging.KeyValueLogMessage;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.record.metadata.MetaDataException;
+import org.apache.lucene.analysis.CharArraySet;
+import org.apache.lucene.analysis.en.EnglishAnalyzer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Registry for {@link SynonymAnalyzer}s.
+ */
+public class SynonymAnalyzerRegistryImpl implements SynonymAnalyzerRegistry {
+    @Nonnull
+    private static final Logger LOGGER = LoggerFactory.getLogger(SynonymAnalyzerRegistryImpl.class);
+
+    private static final CharArraySet STOP_WORDS = EnglishAnalyzer.ENGLISH_STOP_WORDS_SET;
+
+    private static final SynonymAnalyzerRegistryImpl INSTANCE = new SynonymAnalyzerRegistryImpl();
+
+    @Nonnull
+    private Map<String, SynonymAnalyzer> registry;
+
+    private SynonymAnalyzerRegistryImpl() {
+        registry = initRegistry();
+    }
+
+    @Nonnull
+    public static SynonymAnalyzerRegistry instance() {
+        return INSTANCE;
+    }
+
+    @Nonnull
+    private static Map<String, SynonymAnalyzer> initRegistry() {
```

Comment:
I noticed that this implementation doesn't use the `ServiceLoader` like some of our other registries. Is the idea that we'll use a `ServiceLoader` to initialize the registry at some point in the future (but not now)?

Reply:
I was wondering about that. 
Given that we may have hundreds of SynonymAnalyzer which only differ in their constructor parameters, is `ServiceLoader` a good way to approach that? It feels weird to me to define a class for `USEnglishSynonymAnalyzer`, `UKEnglishSynonymAnalyzer`, `DutchSynonymAnalyzer`, etc.
Is there a better way to do this?

Reply:
You could imagine a `ServiceLoader` loadable class that was responsible for contributing more than one analyzer, by loading a JSON file of their locale+location or searching a known file system directory for specific filename patterns or something like that.

Reply:
If we want to eventually support more analyzers by using a ```ServiceLoader``` or some similar thing, how about just loading this default analyzer by that loader?

Reply:
If we want to eventually support more analyzers by using a ServiceLoader or some similar thing, how about just loading this default analyzer by that loader?

Reply:
If we want to eventually support more analyzers by using a ServiceLoader or some similar thing, how about just loading this default analyzer by that loader?

Reply:
Done :+1: